src/std/misc/list.ss

```diff
@@ -304,3 +311,69 @@
     ([] lst)
     ([a] [[a]])
     (_ (helper))))
+
+;; Returns a boolean that is true if any two consecutive terms in the list satisfy the predicate.
+;; In particular, if the predicate is a partial order predicate (respectively a strict partial
+;; order predicate), then the list is totally ordered (respectively strictly totally ordered)
+;; according to the predicate.
+;; (every-consecutive? < [1 2 3 4 5])
+;; => #t
+(def (every-consecutive? f l)
+  (or (null? l)
+      (let loop ((x (car l)) (r (cdr l)))
+        (match r ([] #t) ([y . rr] (and (f x y) (loop y rr)))))))
+
+;; The plist functions below are patterned after pgetq and friends from gerbil/runtime/gx-gambc0.scm
+(defrule (define-pset pset cmp)
+  (def (pset key lst val)
+     (let lp ((rest lst) (rhd []))
+       (match rest
+         ([k v . rest]
+          (if (cmp k key)
+            (foldl cons (cons* key val rest) rhd)
+            (lp rest (cons* v k rhd))))
+         (else
+          (cons* key val lst))))))
+
+(define-pset psetq eq?)
+(define-pset psetv eqv?)
+(define-pset pset equal?)
+
+(defrule (define-pset! pset! cmp)
+  (def (pset! key lst val)
+    (when (null? lst) (error "Cannot destructively modify an empty plist" pset! key lst val))
+    (let lp ((rest lst))
+      (match rest
+        ([k v . more]
+         (if (cmp k key) (set-car! (cdr rest) val) (lp more)))
+        (else
+         (match lst
+           ([k v . rest]
+            (set-car! lst key)
+            (set-cdr! lst (cons* val k v rest)))))))))
+
+(define-pset! psetq! eq?) (def pgetq-set! psetq!)
+(define-pset! psetv! eqv?) (def pgetv-set! psetv!)
+(define-pset! pset! equal?) (def pget-set! pset!)
+
+(defrule (define-prem prem cmp)
+  (def (prem key lst)
+     (let lp ((rest lst) (rhd []))
+       (match rest
+         ([k v . rest]
+          (if (cmp k key)
+            (foldl cons rest rhd)
+            (lp rest (cons* v k rhd))))
+         (else
+          lst)))))
+
+(define-prem premq eq?)
+(define-prem premv eqv?)
+(define-prem prem equal?)
+
+(def (separate-keyword-arguments args)
+  (let lp ((rest args) (positionals []) (keywords []))
+    (match rest
+      ([(? keyword? k) v . r] (lp r positionals (cons* v k keywords)))
+      ([a . r] (lp r (cons a positionals) keywords))
+      ([] (values (reverse positionals) (reverse keywords))))))
```

Comment:
This needs to handle the special `#!key` and `#!rest` constructs.
When you encounter `#!key`, the next argument is a keyword that should be treated as a positional argument.
When you encounter `#!rest`, all the remaining arguments should be treated as positional arguments.

Reply:
See also `keyowrd-dispatch` in `gx-gambc0` about the exact semantics of keyword dispatching.

Reply:
Is it better now?